storage.py

```diff
@@ -0,0 +1,235 @@
+import contextlib
+
+from pymemcache.client.base import Client
+from pymemcache.serde import (python_memcache_serializer,
+                              python_memcache_deserializer)
+
+from clcache import CacheFileStrategy, getStringHash, printTraceStatement, CompilerArtifacts, \
+    CACHE_COMPILER_OUTPUT_STORAGE_CODEC
+
+
+class CacheDummyLock(object):
+    def __enter__(self):
+        pass
+
+    def __exit__(self, typ, value, traceback):
+        pass
+
+
+class CacheMemcacheStrategy(object):
+    def __init__(self, server, cacheDirectory=None, manifestPrefix='manifests_', objectPrefix='objects_'):
+        self.fileStrategy = CacheFileStrategy(cacheDirectory=cacheDirectory)
+        # XX Memcache Strategy should be independent
+
+        self.lock = CacheDummyLock()
+        self.localCache = {}
+        self.localManifest = {}
+        self.objectPrefix = objectPrefix
+        self.manifestPrefix = manifestPrefix
+
+        self.connect(server)
+
+    def connect(self, server):
+        server = CacheMemcacheStrategy.splitHosts(server)
+        assert server, "{} is not a suitable server".format(server)
+        if len(server) == 1:
+            clientClass = Client
+            server = server[0]
+        else:
+            from pymemcache.client.hash import HashClient
+            clientClass = HashClient
+            server = server
+        self.client = clientClass(server, ignore_exc=True,
+                                  serializer=python_memcache_serializer,
+                                  deserializer=python_memcache_deserializer,
+                                  timeout=5,
+                                  connect_timeout=5,
+                                  key_prefix=(getStringHash(self.fileStrategy.dir) + "_").encode("UTF-8")
+                                 )
+        # XX key_prefix ties fileStrategy cache to memcache entry
+        # because tests currently the integration tests use this to start with clean cache
+        # Prevents from having cache hits in when code base is in different locations
+        # adding code to production just for testing purposes
```

Comment:
my concern is that we implement it this way just for testing, to enable
tempfile.TemporaryDirectory() as tempDir:
customEnv = dict(os.environ, CLCACHE_DIR=tempDir)
subprocess.check_call(baseCmd + ["fibonacci01.cpp", "fibonacci02.cpp"], env=customEnv)